File: services/matching-engine/internals/utils/sentry.go
package utils

import (
	"os"

	"github.com/getsentry/sentry-go"
	"github.com/rs/zerolog/log"
)

func SetupSentry() {
	appEnv := os.Getenv("APP_ENV")

	if appEnv == "" {
		appEnv = os.Getenv("NODE_ENV")
	}
	dsn := os.Getenv("SENTRY_DSN")

	if appEnv == "development" || appEnv == "dev" || appEnv == "" || dsn == "" {
		log.Info().Msg("Sentry is disabled (development mode or no DSN)")
		return
	}

	err := sentry.Init(sentry.ClientOptions{
		Dsn:         dsn,
		Environment: appEnv,
		Release:     os.Getenv("SENTRY_RELEASE"),
	})

	if err != nil {
		log.Error().Err(err).Msg("sentry.Init failed")
	} else {
		log.Info().Msg("Sentry initialized")
	}
}

func CaptureError(err error, tags map[string]string, contexts map[string]map[string]interface{}) {
	appEnv := os.Getenv("APP_ENV")

	if appEnv == "" {
		appEnv = os.Getenv("NODE_ENV")
	}

	dsn := os.Getenv("SENTRY_DSN")

	if appEnv == "development" || appEnv == "dev" || appEnv == "" || dsn == "" {
		return
	}

	sentry.WithScope(func(scope *sentry.Scope) {
		if tags != nil {
			scope.SetTags(tags)
		}

		for key, val := range contexts {
			scope.SetContext(key, sentry.Context(val))
		}
		sentry.CaptureException(err)
	})
}

File: services/stream-service/src/libs/sentry.ts
import { ENV } from '@/config/env';
import * as Sentry from '@sentry/bun';

export type SentryContext = {
	tags: {
		action: string;
		[key: string]: string | number | boolean | undefined;
	};
	contexts?: {
		[key: string]: Record<string, any>;
	};
};

export const setupSentry = () => {
	if (ENV.NODE_ENV === 'development' || !ENV.SENTRY_DSN) return;

	Sentry.init({
		dsn: ENV.SENTRY_DSN,
		environment: ENV.NODE_ENV,
		tracesSampleRate: 1.0,
	});
};

export const captureError = (error: unknown, context?: SentryContext | Record<string, any>) => {
	if (ENV.NODE_ENV === 'development' || !ENV.SENTRY_DSN) return;

	Sentry.withScope((scope) => {
		if (context) {
			if ('tags' in context || 'contexts' in context) {
				const typedContext = context as SentryContext;
				if (typedContext.tags) {
					scope.setTags(typedContext.tags);
				}
				if (typedContext.contexts) {
					Object.entries(typedContext.contexts).forEach(([key, value]) => {
						scope.setContext(key, value);
					});
				}
			} else {
				scope.setTags(context as Record<string, string>);
			}
		}

		if (error instanceof Error) {
			Sentry.captureException(error);
		} else {
			Sentry.captureMessage(String(error), 'error');
		}
	});
};
